### PR TITLE
Adding automargin support to plot titles

### PR DESCRIPTION
**Pull request by hannahker**
This pull request adds 'automargin' support for `layout.title` components. If `layout.title.automargin=true`, then either the top or the bottom margin space will expand such that the plot title is visible in the plotting area. 

**To-dos:** 
- [x] Add hard-coded support for `title.automargin`
- [x] Fix failing tests for pie charts 
- [x] Determine whether margin should be added to top or bottom based on `layout.title.y`
- [x] Push margin depending on title font size
- [x] Set `title.y=1` and `title.yref='paper'` defaults when `title.automargin=true` (in `supplyDefaults` stage)
- [x] Allow padding to be added to both top and bottom, irrespective of `yanchor`
- [x] Have title avoid overlap with container when `yref='paper'`
- [x] Write Jasmine tests for `yref='paper'`
- [x] Have title avoid overlap with plot area when `yref='container'`
- [x] Write Jasmine tests for `yref='container'`
- [x] BUG: Fix interactions with automargined x-axes when title is on the bottom of the plot area (and test interactions with other elements adding margin to the top)
- [ ] NA: Set defaults in defaults setting phase
- [x] OUT OF SCOPE: Account for multi-line titles (not just estimation based on font size) (maybe relevant https://github.com/plotly/plotly.js/issues/4299) -- also see https://github.com/plotly/plotly.js/issues/6472
- [x] BUG: Fix errors in unnecessary redrawing when `yref='paper'` 

**API Questions:** 
- [x] Should `title.automargin` support multi-line titles? Looking at [this codepen](https://codepen.io/hannahker/pen/gOjyLEY), there seem to be some issues with multi-line positioning, see https://github.com/plotly/plotly.js/issues/6472
  **A: Currently out of scope given existing bug.**
 
- [x] Should `title.automargin` allow the title to overlap with the plot itself? Or does `title.automargin` only ensure that the title is visible? Ex. what if `title.y=0.6`? Should the margin then expand to take up the top 40% of the plot area?
     **A: This depends on `title.yref`. When the title is `'paper'` referenced, then `title.automargin` can't do anything to impact potential overlap with the plot area, as the plot area is also `'paper'` referenced. When the title is `'container'` referenced, then `title.automargin` will increase the top or bottom margin to avoid overlap with the plot area.**

- [x] If a developer manually specifies `title.yref='container'` and `title.automargin`, then which one 'wins'?
     **A: N/A. In this case `title.automargin` will have an impact.** 

- [x] `layout.title.y=1` + `layout.title.yanchor='bottom'` / `layout.title.y=0` + `layout.title.yanchor='top` --> should `automargin` handle these cases so that the title is visible? Logically, I don't think it should be. 
     **A: Assuming that `title.yref='container'`, the title should not be visible in these cases as it is outside of the plot area. If `title.yref='paper'`, then this would be a common way to pin the title to either the top or bottom of the plot area.** 

- [x] What if `layout.title.y='auto'` and there isn't enough manually-defined top margin to fit the title? (ref: ""auto" places the baseline of the title onto the vertical center of the top margin"). Should we add in logic here so that the top margin will expand to 2x the size of the title? It doesn't seem intuitive to the developer in this case that the top margin will have this artificial padding. 
     **A: When `title.automargin=true`, we can set the default `title.y=1`, which is a more logical default and would avoid this behaviour.** 

- [x] How will `title.automargin` behaviour be different when `yref='container'` vs `yref='paper'`?
     **A: If `yref='container'`, then `title.automargin` will push margins to ensure that the title doesn't overlap with the plot area.  If `yref='paper'`, then `title.automargin` will only push margins to ensure that the title doesn't overlap with the edges of the container.**

- [x] Should `title.automargin` potentially push out the left or right margins?
     **A: No, this implementation will only support spacing added to top or bottom margins.**

**Technical Questions:**
- [x] How should the `x` and `y` input parameters to the `Plots.autoMargin` function be specified? I would expect that these would be just `title.x` and `title.y`. In testing, seems only to work when `y` is `1` or `0`. (ref: "x, y: plot fraction of the anchor point.")
- [x] How do different automargined features interact with each other when they're pushing out the same margin? Ex. when `title.automargin` applies bottom margin as well as the x-axis?

### CODE REVIEW

File: src/plot_api/subroutines.js [hunk 1]
```diff
@@ -438,14 +514,7 @@ function getMainTitleX(fullLayout, textAnchor) {
...
     var gs = fullLayout._size;
-    var vPadShift = 0;
-
-    if(dy === '0em' || !dy) {
-        vPadShift = -title.pad.b;
-    } else if(dy === alignmentConstants.CAP_SHIFT + 'em') {
-        vPadShift = title.pad.t;
-    }
-
+    var vPadShift = getVPadShift(title, dy);
```
Comment (archmoj) [inline src/plot_api/subroutines.js:517]: Do we have to declare `getVPadShift` function above and call it here?
It looks like it is not get called from other places.
So perhaps you could remove it to revert the changes here.

File: src/plot_api/subroutines.js [hunk 2]
```diff
@@ -397,24 +399,98 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
+    var title = gd._fullLayout.title;
+    setDflts(title);
     var fullLayout = gd._fullLayout;
-
     var textAnchor = getMainTitleTextAnchor(fullLayout);
     var dy = getMainTitleDy(fullLayout);
+    var y = getMainTitleY(fullLayout, dy);
+    var position = fullLayout.title.y > 0.5 ? 't' : 'b';
+
+    if(gd._fullLayout.title.automargin) {
```
Comment (archmoj) [inline src/plot_api/subroutines.js:410]: `title` is declared here. So we can rewrite:
```suggestion
    if(title.automargin) {
```

File: src/plot_api/subroutines.js [hunk 3]
```diff
@@ -397,24 +399,98 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
 
 exports.drawMainTitle = function(gd) {
+    var title = gd._fullLayout.title;
+    setDflts(title);
     var fullLayout = gd._fullLayout;
-
     var textAnchor = getMainTitleTextAnchor(fullLayout);
     var dy = getMainTitleDy(fullLayout);
+    var y = getMainTitleY(fullLayout, dy);
+    var position = fullLayout.title.y > 0.5 ? 't' : 'b';
```
Comment (archmoj) [inline src/plot_api/subroutines.js:408]: This `position` variable is only used inside `applyTitleAutoMargin`. 
Let's remove it from here and put it in `applyTitleAutoMargin` function which would reduce the number of arguments passed to it too.

File: src/plot_api/subroutines.js [hunk 4]
```diff
@@ -397,24 +399,98 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
+        fontSize * (extraLines + 1) * LINE_SPACING :
+        fontSize;
+}
+
+function applyTitleAutoMargin(gd, position, y) {
+    var titleID = 'title.automargin';
+    var title = gd._fullLayout.title;
+    var push = {
+        x: title.x,
+        y: position === 't' ? 1 : 0, // TODO: Assuming that y is either 1 or 0?
```
Comment (archmoj) [inline src/plot_api/subroutines.js:468]:
```suggestion
        y: title.y > 0.5 ? 1 : 0,
```

File: src/plot_api/subroutines.js [hunk 5]
```diff
@@ -397,24 +399,98 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
+// Is any part of the title outside of the container?
+// TODO: What if yref='container'?
+function isOutsideContainer(gd, title, position, y) {
+    var plotHeight = gd._fullLayout._size.h;
+    var yPosTop = Lib.isTopAnchor(title) ? y : y - titleDepth(title); // Standardize to the top of the title
+    var yPosRel = position === 'b' ? plotHeight - yPosTop : yPosTop; // Position relative to the top or bottom of plot
+    if((Lib.isTopAnchor(title) && position === 't') || Lib.isBottomAnchor(title) && position === 'b') {
+        return false;
+    } else {
+        if(yPosRel < titleDepth(title)) {
```
Comment (archmoj) [inline src/plot_api/subroutines.js:436]: We could simplfiy these last two blocks here:
```js
        return yPosRel < titleDepth(title);
```

File: src/plot_api/subroutines.js [hunk 6]
```diff
@@ -397,24 +399,98 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
+            return true;
+        } else {
+            return false;
+        }
+    }
+}
+
+// TODO: Move to setting defaults stage, rather than drawing
+// title.y is 1 or 0 if automargin and paper ref
```
Comment (archmoj) [inline src/plot_api/subroutines.js:445]: Yes we should move this part to supply defaults step.
I feel the `y` position should not be defined when `automargin` is *true*.
And instead of `y` values to position the title on top or at the bottom, we could/should use `yanchor` values.
Does this make sense?
Also let's add a separate mock to test the positioning of the title at the bottom.
Thank you!

**Review by archmoj — COMMENTED** [review]
@hannahker very nice improvement.
Please find my comments below.
Thank you!

File: src/plot_api/subroutines.js (hunk 6 above)
Comment (hannahker) [reply src/plot_api/subroutines.js:441]: Yes, I'll add in the mock! Using `yanchor` to control top of bottom position would be interesting... Although I think this would add complexity for the `container` positioning as then the `y` is relevant for pushing the size of the margin (something that I also need to add in tests around).

File: test/image/mocks/zzz-automargin-title-container.json [hunk 7]
```diff
@@ -18,11 +18,14 @@
         "height": 300,
         "width": 400,
         "margin": {"t":0, "b": 0, "l": 0, "r": 0},
+        "xaxis": {"automargin": true, "title": {"text": "x-axis title", "standoff": 0}},
```
Comment (hannahker) [inline test/image/mocks/zzz-automargin-title-container.json:21]: @alexcjohnson - The bottom margin ends up being set to the correct size to account for both the plot title and the axis, but the axis title then still moves to overlap with the plot title when given enough margin space. Manually setting the `standoff` to `0` fixes this, but I can't quite pinpoint why this is the case (as I don't see the margin size or plot height as inputs to the automatic standoff calculation in the code. Do you have any thoughts on where this might be coming from?

File: src/plot_api/plot_api.js [hunk 8]
```diff
@@ -277,6 +277,7 @@ function _doPlot(gd, data, layout, config) {
 
         subroutines.drawMarginPushers(gd);
         Axes.allowAutoMargin(gd);
+        Plots.allowAutoMargin(gd, 'title.automargin');
```
Comment (archmoj) [inline src/plot_api/plot_api.js:280]: Shouldn't we ensure there is a title text and automargin before calling this?
```suggestion
        if(gd._fullLayout.title.text && gd._fullLayout.title.automargin) Plots.allowAutoMargin(gd, 'title.automargin');
```

File: src/plot_api/subroutines.js [hunk 9]
```diff
@@ -397,24 +399,128 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
 exports.drawMainTitle = function(gd) {
+    var title = gd._fullLayout.title;
+    setDflts(title);
     var fullLayout = gd._fullLayout;
-
     var textAnchor = getMainTitleTextAnchor(fullLayout);
     var dy = getMainTitleDy(fullLayout);
+    var y = getMainTitleY(fullLayout, dy);
+
+    if(title.automargin) {
```
Comment (archmoj) [inline src/plot_api/subroutines.js:409]: Possibly
```suggestion
    if(title.text && title.automargin) {
```

File: src/plot_api/subroutines.js [hunk 10]
```diff
@@ -397,24 +399,128 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
+// 'auto' is not supported for either title.y or title.yanchor when automargin=true
+function setDflts(title) {
+    if(title.automargin && title.yref === 'paper') {
+        title.y = title.y === 0 ? title.y : 1;
+        if(title.yanchor === 'auto') {
+            if(title.y === 0) {
+                title.yanchor = 'top';
+            } else {
+                title.yanchor = 'bottom';
+            }
```
Comment (archmoj) [inline src/plot_api/subroutines.js:451]: Or simply
```js
title.yanchor = title.y === 0 ? 'top' : 'bottom';
```

File: src/plot_api/subroutines.js [hunk 11]
```diff
@@ -397,24 +399,128 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
+        }
+    }
+    if(title.automargin && title.yref === 'container') {
+        title.y = title.y === 'auto' ? 1 : title.y;
```
Comment (archmoj) [inline src/plot_api/subroutines.js:455]:
```suggestion
        if(title.y === 'auto') title.y = 1;
```

File: src/plots/layout_attributes.js [hunk 12]
```diff
@@ -125,6 +125,14 @@ module.exports = {
                 'Padding is muted if the respective anchor value is *middle*/*center*.'
             ].join(' ')
         }),
+        automargin: {
+            valType: 'boolean',
+            dflt: false,
+            editType: 'plot',
+            description: [
+                'Determines whether the title can automatically push the figure margins.'
```
Comment (archmoj) [inline src/plots/layout_attributes.js:133]: Could you please improve the description to mention the role of other attributes here?
Thank you!

File: test/jasmine/tests/titles_test.js [hunk 13]
```diff
@@ -1066,6 +1066,97 @@ describe('Titles for multiple axes', function() {
     });
 });
 
+// TODO: Add in tests for interactions with other automargined elements
```
Comment (archmoj) [inline test/jasmine/tests/titles_test.js:1069]: Could we drop this TODO now?

File: src/plots/plots.js [hunk 14]
```diff
@@ -2016,14 +2026,16 @@ plots.doAutoMargin = function(gd) {
             var pl = pushleft.size;
             var fb = pushbottom.val;
             var pb = pushbottom.size;
+            var nonReservedWidth = width - reservedMargins.r - reservedMargins.l;
```
Comment (archmoj) [inline src/plots/plots.js:2029]: `availableWidth` may read better. No?

File: src/plots/plots.js [hunk 15]
```diff
@@ -1979,8 +1981,16 @@ plots.doAutoMargin = function(gd) {
...
+    var reservedMargins = {'t': 0, 'b': 0, 'l': 0, 'r': 0};
     var oldMargins = Lib.extendFlat({}, gs);
 
+    // We're assuming that each entry only pushes the margin out on one side
+    var margins = gd._fullLayout._reservedMargin;
+    for(var key in margins) {
+        var side = Object.keys(margins[key])[0];
+        var val = Object.values(margins[key])[0];
+        reservedMargins[side] += val;
+    }
```
Comment (alexcjohnson) [inline src/plots/plots.js:1993]:
```suggestion
    var margins = gd._fullLayout._reservedMargin;
    for(var key in margins) {
        for(var side in margins[key]) {
            var val = margins[key][side];
            reservedMargins[side] = Math.max(reservedMargins[side], val);
        }
    }
```
- No reason to impose the limitation of only one side - I could easily see an object pushing out both top and right, for example. And if there's only one, adding this loop is no slower, might even be faster.
- Right now you only have one so this doesn't matter, but when there are multiple if we want them to push each other around (and sum their reserved pixels) we'll need some other mechanism. Until then they won't add, you'll just take the largest one. For example you might put a title and legend both container-positioned on top, with the title on the left and the legend on the right.

File: src/plots/plots.js [hunk 16]
```diff
@@ -1979,8 +1981,16 @@ plots.doAutoMargin = function(gd) {
 
     var gs = fullLayout._size;
     var margin = fullLayout.margin;
+    var reservedMargins = {'t': 0, 'b': 0, 'l': 0, 'r': 0};
```
Comment (alexcjohnson) [inline src/plots/plots.js:1984]:
```suggestion
    var reservedMargins = {t: 0, b: 0, l: 0, r: 0};
```
we don't have a linter that enforces this?

**Comment by archmoj** [comment]
Nice work @hannahker :medal_military: 
It's getting close.
On a dev branch I tested rendering of all of our mocks by setting automargin to true.
They do render without error which is great. 
However I noticed the positioning appears to be slightly off.
Below is one example showing mock labeled `4`.

diff-mock-4

File: src/plot_api/subroutines.js (hunk 6 above)
Comment (alexcjohnson) [reply src/plot_api/subroutines.js:441]: Yes, this logic should all move into supplyDefaults.

I still don't understand why we're restricting `title.y` to either 0 or 1. Yes, 99% of the time that's what people will use, but if someone gets into the details of tweaking their graph and wants to use `title.y` to move the title a little up or down why would we prevent this? Worse still, if they've forgotten that they turned on `title.automargin` and they try to adjust `title.y` from 0 to 0.01 and it snaps all the way to the top, they'll have to somehow figure out that they need to turn off `title.automargin` before they can get the desired effect?

File: test/image/mocks/zzz-automargin-title-container.json (hunk 7 above)
Comment (alexcjohnson) [reply test/image/mocks/zzz-automargin-title-container.json:21]: Looks to me like it's probably this block, which creates a default standoff that's not recorded anywhere or captured anywhere else, just fed into where the axis title is drawn:

https://github.com/plotly/plotly.js/blob/2b4584dfe54598db03e8d9724a766789293b4f69/src/plots/cartesian/axes.js#L3973-L3998

The ideal solution would be put all that logic into `title.standoff` in `supplyDefaults`, rather than at the draw step.

BTW @archmoj I notice `standoff` is incorrect for `multicategory` axes with `side: 'top'`... didn't look at left or right but I bet one of those is incorrect too.

File: src/plots/plots.js (hunk 16 above)
Comment (archmoj) [reply src/plots/plots.js:1986]: Enforced by #6513.

File: src/plots/plots.js (hunk 14 above)
Comment (hannahker) [reply src/plots/plots.js]: Agreed!

File: src/plots/plots.js (hunk 15 above)
Comment (hannahker) [reply src/plots/plots.js]: Agreed, this makes sense!

File: src/plots/plots.js (hunk 16 above)
Comment (hannahker) [reply src/plots/plots.js]: Amazing! I made this mistake on a previous PR and it left me stuck for a while...

**Comment by hannahker** [comment]
@archmoj this behaviour is coming from two places: 
1. The title moves to the top because when `title.automargin=true` the previous default `title.y='auto'` is replaced by `title.y=1`. 
2. The top margin also unnecessarily adds the title's height (17px) to the existing 60px margin already defined. 

Looks like this is pointing to the need to better handle cases where `title.automargin=true` and `title.yref='container'`, but the automargining doesn't need to do anything. Point #1 seems like a case to me where we actually do want the `setDflts()` function to remain in `subroutines.js`, where we can identify the condition that the user-defined margin is already large enough to accommodate for the title and so we don't actually need to adjust `title.y='auto'`. 

@archmoj after applying the review comments from yourself and @alexcjohnson I've gotten the same failing pie chart test cases that you solved previously in these commits -- [here](https://github.com/plotly/plotly.js/pull/6428/commits/f46014fbd8297e4d717db9903a1d6daf78e2f040),  [here](https://github.com/plotly/plotly.js/pull/6428/commits/587dfcbe389a416e46f28ae0c27451454b1b9645), and [here](https://github.com/plotly/plotly.js/pull/6428/commits/3dae10955bf5da8fa8a8fb65ce48c040853ad7a7). I still don't entirely understand what's going on with these failures and how consequential they are. Do you have any thoughts?

File: test/image/mocks/zzz-automargin-title-container.json (hunk 7 above)
Comment (hannahker) [reply test/image/mocks/zzz-automargin-title-container.json:21]: I was looking into this block as well but it's not clear to me how a change in margin size impacts the standoff value. By my reading, it's only the font size and axis properties (like tick length and line width) that we're looking at here. I could be missing something obvious since I've been staring at this block for a while.

**Comment by archmoj** [comment]
> @archmoj after applying the review comments from yourself and @alexcjohnson I've gotten the same failing pie chart test cases that you solved previously in these commits -- [here](https://github.com/plotly/plotly.js/pull/6428/commits/f46014fbd8297e4d717db9903a1d6daf78e2f040), [here](https://github.com/plotly/plotly.js/pull/6428/commits/587dfcbe389a416e46f28ae0c27451454b1b9645), and [here](https://github.com/plotly/plotly.js/pull/6428/commits/3dae10955bf5da8fa8a8fb65ce48c040853ad7a7). I still don't entirely understand what's going on with these failures and how consequential they are. Do you have any thoughts?

Excellent. So now  you could simply revert https://github.com/plotly/plotly.js/pull/6428/commits/3dae10955bf5da8fa8a8fb65ce48c040853ad7a7 :crossed_fingers:

**Comment by archmoj** [comment]
> @archmoj this behaviour is coming from two places:
> 
> 1. The title moves to the top because when `title.automargin=true` the previous default `title.y='auto'` is replaced by `title.y=1`.
> 2. The top margin also unnecessarily adds the title's height (17px) to the existing 60px margin already defined.
> 
> Looks like this is pointing to the need to better handle cases where `title.automargin=true` and `title.yref='container'`, but the automargining doesn't need to do anything. Point #1 seems like a case to me where we actually do want the `setDflts()` function to remain in `subroutines.js`, where we can identify the condition that the user-defined margin is already large enough to accommodate for the title and so we don't actually need to adjust `title.y='auto'`.
> 

Thanks for the explanation. It would be great to improve the documentation for the new attribute in this regard.

File: src/plot_api/subroutines.js (hunk 6 above)
Comment (hannahker) [reply src/plot_api/subroutines.js:441]: Agreeing to leave `title.y` as 0 or 1 in this case, and add a console warning when it is being reverted to these values.

File: test/image/mocks/zzz-automargin-title-container.json (hunk 7 above)
Comment (hannahker) [reply test/image/mocks/zzz-automargin-title-container.json:21]: Look to `scootTitle` function in main `Titles.draw`

File: src/plots/layout_attributes.js (hunk 12 above)
Comment (hannahker) [reply src/plots/layout_attributes.js:133]: How do we feel about this: 

 > Determines whether the title can automatically push the figure margins. If yref=paper then the margin will expand to ensure that the title doesn’t overlap with the edges of the container. If yref=container then the margins will also ensure that the title doesn’t overlap with the plot area. If automargin=true and the margins need to be expanded, then y will be set to a default 1 and yanchor will be set to an appropriate default to ensure that minimal margin space is needed. Note that when yref=paper, only 1 or 0 are allowed y values. Invalid values will be reset to the default 1. 

cc @alexcjohnson

**Comment by hannahker** [comment]
@archmoj see [this commit ](https://github.com/plotly/plotly.js/pull/6428/commits/17c32447978d67ecd835ee017e091bbfeb586a1c) for updates that should ensure that `automargin` doesn't change the title properties unless required. Because of these changes, I don't think it's reasonable to put the defaults setting into `supplyDefaults` .

File: src/plot_api/subroutines.js [hunk 17]
```diff
@@ -463,6 +463,9 @@ function getDflts(title) {
 }
 
 function setDflts(title, titleY, titleYanchor) {
+    if(title.yref === 'paper' && title.y !== 0 && title.y !== 1) {
+        console.warn("title.automargin=true so resetting the supplied title.y value to 1.")
```
Comment (hannahker) [inline src/plot_api/subroutines.js]: I can see that this is giving a linting error. @archmoj is there a better way to raise this?

Comment (archmoj) [reply src/plot_api/subroutines.js]: Looking good to me.

File: src/components/titles/index.js [hunk 18]
```diff
@@ -167,11 +167,21 @@ function draw(gd, titleClass, options) {
...
+                    reservedMargins[side] = Math.max(reservedMargins[side], val);
+                }
+            }
             var paperbb = {
-                left: 0,
-                top: 0,
-                right: fullLayout.width,
-                bottom: fullLayout.height
+                left: 0 + reservedMargins.l,
+                top: 0 + reservedMargins.t,
```
Comment (archmoj) [inline src/components/titles/index.js:182]:
```suggestion
                left: reservedMargins.l,
                top: reservedMargins.t,
```

File: src/plot_api/subroutines.js (hunk 17 above)
Comment (hannahker) [reply src/plot_api/subroutines.js]: How should I get around the syntax test failure?

File: src/plot_api/subroutines.js [hunk 19]
```diff
@@ -397,24 +399,151 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
+
+function setDflts(title, titleY, titleYanchor) {
+    if(title.yref === 'paper' && title.y !== 0 && title.y !== 1) {
+        console.warn('title.automargin=true so resetting the supplied title.y value to 1.');
+    }
+    title.y = titleY;
+    title.yanchor = titleYanchor;
+}
+
+function titleDepth(title) {
```
[the remark this answers is not in the document]
Comment (emilykl) [reply src/plot_api/subroutines.js:473]: @archmoj @hannahker This is the relevant code for shape labels... happy to discuss if that would be helpful.

https://github.com/plotly/plotly.js/blob/ea8e6d11e571f9c2de7ac880bafb97e59a345adb/src/components/shapes/draw.js#L682

[the remark this answers is not in the document]
Comment (hannahker) [reply src/plot_api/subroutines.js:473]: @archmoj I had something like this figured out, although was struggling to reference this appropriately since I need the title to be drawn first. What would be the best way to draw the title 'offscreen' as you had suggested? 

```
var titleHeight = Drawing.bBox(d3.selectAll('.gtitle').node()).height; 
```

File: src/components/titles/index.js [hunk 20]
```diff
@@ -167,11 +167,21 @@ function draw(gd, titleClass, options) {
             var pad = isNumeric(avoid.pad) ? avoid.pad : 2;
 
             var titlebb = Drawing.bBox(titleGroup.node());
+
+            // Need to account for reservedMargins here as well
```
[the remark this answers is not in the document]
Comment (hannahker) [reply src/components/titles/index.js]: Yes, correct

File: src/plot_api/subroutines.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (archmoj) [reply src/plot_api/subroutines.js]: > 

@hannahker Please push your WIP into a dev branch so that we could look at it

[the remark this answers is not in the document]
Comment (hannahker) [reply src/plot_api/subroutines.js]: See [here](https://github.com/plotly/plotly.js/commit/e77126c2efec72e605e6e8eea9192a79844843ea)!

[the remark this answers is not in the document]
Comment (hannahker) [reply src/plot_api/subroutines.js]: This seems to work well, although I'm still noticing that the drawn title position is offset with some of the bounding box outside of the container. This isn't an issue of margins, but rather I think the `dy` attribute (see setting [here](https://github.com/plotly/plotly.js/blob/master/src/plot_api/subroutines.js#L475)). I can see this behaviour on the `master` branch as well so this might be an existing bug. 

Screen Shot 2023-03-10 at 10 29 10 AM

[the remark this answers is not in the document]
Comment (archmoj) [reply src/plot_api/subroutines.js]: I agree with you that this function 
https://github.com/plotly/plotly.js/blob/ea8e6d11e571f9c2de7ac880bafb97e59a345adb/src/plot_api/subroutines.js#L475-L486
could be the source of bug on master.
Let's try not to use it when `automargin` is set to true.

File: src/plots/layout_attributes.js (hunk 12 above)
Comment (alexcjohnson) [reply src/plots/layout_attributes.js:133]: Excellent! Just a small edit in the 3rd sentence, did I get this right?

> Determines whether the title can automatically push the figure margins. If yref=paper then the margin will expand to ensure that the title doesn’t overlap with the edges of the container. If yref=container then the margins will ensure that the title doesn’t overlap with the plot area, tick labels, and axis titles. If automargin=true and the margins need to be expanded, then y will be set to a default 1 and yanchor will be set to an appropriate default to ensure that minimal margin space is needed. Note that when yref=paper, only 1 or 0 are allowed y values. Invalid values will be reset to the default 1.

Comment (hannahker) [reply src/plots/layout_attributes.js:133]: Yes, thanks for adding that!

File: src/plot_api/subroutines.js [hunk 21]
```diff
@@ -397,24 +399,156 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
+    return [titleY, titleYanchor];
+}
+
+function setDflts(title, titleY, titleYanchor) {
+    if(title.yref === 'paper' && title.y !== 0 && title.y !== 1 && title.y !== 'auto') {
+        Loggers.warn('title.automargin=true so resetting the supplied title.y value to 1.');
+    }
+    title.y = titleY;
+    title.yanchor = titleYanchor;
+}
```
Comment (archmoj) [inline src/plot_api/subroutines.js:483]: @hannahker wondering if you may possibly consider picking this commit: https://github.com/plotly/plotly.js/commit/0a23e1d19bab780aad6a1941d5d281cb31c0fe4c
which moves this logic to the supply defaults step?

Comment (hannahker) [reply src/plot_api/subroutines.js:483]: Thanks @archmoj! [This piece](https://github.com/plotly/plotly.js/pull/6428/files#diff-d670f85239a0f2240b21379a074593658aa18b142d90eeeab8a10699713eb148R426) is what's making me think that we can't move the default setting to the supply defaults step, as we'll only do it if the margin actually needs to be pushed (addressing the scenario in your screenshots above)

File: test/image/mocks/zzz-automargin-title-container.json [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+        "height": 300,
+        "width": 400,
+        "margin": {"t":0, "b": 0, "l": 0, "r": 0},
+        "xaxis": {"automargin": true, "title": {"text": "x-axis title"}},
+        "title": {
+            "automargin": true, 
+            "text": "Basic title",
+            "font": {"size": 24},
+            "pad": {"t": 0, "b": 0},
+            "y": 0,
```
Comment (archmoj) [inline test/image/mocks/zzz-automargin-title-container.json:27]: Please add a new mock exactly like this one; but using `y: 1`.
Thank you!

File: test/image/mocks/zzz-automargin-title-paper.json [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+    "layout": {
+        "height": 300,
+        "width": 400,
+        "margin": {"t":0, "b": 0, "l": 0, "r": 0},
+        "title": {
+            "automargin": true, 
+            "text": "Basic title",
+            "font": {"size": 24},
+            "pad": {"t": 15, "b": 10},
+            "yref": "paper"
```
Comment (archmoj) [inline test/image/mocks/zzz-automargin-title-paper.json:26]: Also please add a new mock exactly like this one; but using `y: 0`.
Thanks again.

File: src/plot_api/subroutines.js [hunk 24]
```diff
@@ -397,24 +399,156 @@ function findCounterAxisLineWidth(ax, side, counterAx, axList) {
...
+    if(position === 't') {
+        if(titleYanchor === 'top') {
+            push += titleDepth;
+        }
+        push += (height - titleY * height);
+    } else {
+        if(titleYanchor === 'bottom') {
+            push += titleDepth;
+        }
+        push += (height - (1 - titleY) * height);
```
Comment (archmoj) [inline src/plot_api/subroutines.js:500]:
```suggestion
        push += titleY * height;
```

File: test/image/mocks/zzz-automargin-title-container.json [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+    "layout": {
+        "height": 300,
+        "width": 400,
+        "margin": {"t":0, "b": 0, "l": 0, "r": 0},
+        "xaxis": {"automargin": true, "title": {"text": "x-axis title"}},
+        "title": {
+            "automargin": true, 
+            "text": "Basic title",
+            "pad": {"t": 0, "b": 0},
+            "y": 0,
```
Comment (archmoj) [inline test/image/mocks/zzz-automargin-title-container.json:26]: So that it would be on top?
```suggestion
            "y": 1,
```

File: test/image/mocks/zzz-automargin-title-container-bottom.json [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+            ]
+        }],
+    "layout": {
+        "height": 300,
+        "width": 400,
+        "margin": {"t":0, "b": 0, "l": 0, "r": 0},
+        "xaxis": {"automargin": true, "title": {"text": "x-axis title"}},
+        "title": {
+            "automargin": true, 
+            "text": "Basic title",
```
Comment (archmoj) [inline test/image/mocks/zzz-automargin-title-container-bottom.json:24]: Thanks @hannahker for adding new tests.
Now let's use `container` and `paper` somewhere within the `title.text` so that the tests become self describing.
I also used a descender like `p` in the text.
```suggestion
            "text": "Container | no-pad",
```

File: test/image/mocks/zzz-automargin-title-paper.json [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+            6
+            ]
+        }],
+    "layout": {
+        "height": 300,
+        "width": 400,
+        "margin": {"t":0, "b": 0, "l": 0, "r": 0},
+        "title": {
+            "automargin": true, 
+            "text": "Paper | no-pad | y=1",
```
Comment (archmoj) [inline test/image/mocks/zzz-automargin-title-paper.json:23]: Would you please add yet another mock with multi-line text (either 2 or 3 lines) and commit the baseline, even if it may not position perfectly? This would help use debug and possibly fix the issue.
Thank you!

**Review by alexcjohnson — APPROVED** [review]
Excellent, I think we finally got there! 💃 
I know there's still an issue with multiline titles - as shown in the `zzz-automargin-title-paper-multiline` mock - but since this was already buggy before this PR we can address that in a patch.

**Comment by Alexboiboi** [comment]
This is so good for multiline titles, thank you very much!
